docs/advanced_topics.rst

```diff
@@ -27,10 +27,15 @@ this will set thread 0 to CPU 3, thread 1 to CPU 5, thread 2 to CPU 6, thread 3
 Memory consumption
 ------------------
 
-In DALI architecture all GPU operators process the whole batch at a time while the CPU operator processes one sample at a time.
-With DALI the memory is only growing, never shrinking. Hence each GPU buffer is as large as the largest possible batch, while the CPU buffers are as large as batch size multiplied by the size of the largest sample. Note that even though the CPU processes one sample at a time per thread, a vector of samples need to reside in the memory.
-This lazy allocation strategy reduces the number of total memory operations, and hence helps in cases when the memory allocation is expensive, like the pinned CPU memory or GPU memory.
-This is most visible for the operators whose output size may differ from sample to sample and from run to run. Operator with the fixed size outputs, such as crop, does not influence the overall memory consumption growth over time
+DALI uses three kinds of memory: host, host page-locked (pinned) and GPU.
+
+GPU buffers are allocated to house transformation results is as large as the largest possible batch, while the CPU buffers can be as large as batch size multiplied by the size of the largest sample. Note that even though the CPU processes one sample at a time per thread, a vector of samples needs to reside in the memory. It is worth to note that some CPU operators can calculate the output shape (and thus, the memory required) ahead of time, in which case the output will be preallocated as a single continuous buffer for the whole batch, which makes their memory consumption on par with their GPU counterparts.
```

Comment:
I think this lacks a bit of context, maybe something like:
```
DALI works on batches of samples. For GPU Operators the batch is stored as continuous allocation which is processed in one go. This again reduces the number of necessary allocations.
For some CPU Operators, that cannot calculate their output size ahead of time, the batch is instead stored as a vector of separately allocated samples (for others it's still a single continuous allocation).

For example if your batches consists of nine 480p images and one 4K image in random order, the single continuous allocation would be able to accommodate all possible combinations of such batches.
On the other hand, the CPU batch presented as separate buffers will need to keep an 4K allocation for every sample after several iterations.
```

The example at the end can be swapped for something less concrete.

Reply:
Done